diagnostic.js

```diff
@@ -32,25 +32,88 @@ const Movement = require('./models/movement.js');
 
 /// ADD YOUR CODE BELOW
 
-const create = (name, description, startYear, endYear) => {};
-// Success -> console.log new Movement as JSON
-// Failure -> Console.error
+const db = mongoose.connection;
 
-const index = () => {};
-// Success -> console.log all Movements as JSON
-// Failure -> Console.error
+const done = function() {
+  db.close();
+};
+
+const create = (name, description, startYear, endYear) => {
+  Movement.create({
+    name: name,
+    description: description,
+    startYear: startYear,
+    endYear: endYear
+  }).then(function(movement) {
+    console.log(movement.toJSON());
+  }).catch(function(error) {
+    console.error(error);
+  }).then(done);
+};
+
+const index = () => {
+  let search = {};
+  if (arguments[0] && arguments[1]) {
+    let field = arguments[0];
+    let criterion = arguments[1];
+    if (criterion[0] === '/') {
+      let regex = new RegExp(criterion.slice(1, criterion.length - 1));
+      search[field] = regex;
+    } else {
+      search[field] = criterion;
+    }
+  }
+  Movement.find(search).then(function(movements) {
+    movements.forEach(function(movement) {
+      console.log(movement.toJSON());
+    });
+  }).catch(console.error).then(done);
+};
+
+const show = (id) => {
+  Movement.findById(id).then(function(movement) {
+    if (movement) {
+      console.log(movement.toObject());
+    }
+    else {
+      return console.error("Not Found");
+    }
+  }).catch(console.error).then(done);
+};
 
-const show = (id) => {};
 // Success -> If the specified Movement exists, console.log it as JSON;
 //              otherwise, console.log "Not Found" and exit.
 // Failure -> Console.error
 
-const update = (id, field, value) => {};
+const update = (id, field, value) => {
+  let modify = {};
+  modify[field] = value;
+  Movement.findById(id).then(function(movement) {
+    if (movement) {
+      movement.set(field, value);
+      return movement.save();
+    }
+    else {
+      return console.error("Not Found");
+    }
+  }).then(function(movement) {
+    console.log(movement.toJSON());
```

Comment:
If a movement isn't found, then the first call to `then()` doesn't return a value.  The second `then()` will have `undefined` as `movement` and invoking `movement.toJSON()` will throw an error.